prow/cmd/peribolos/main.go

```diff
@@ -103,21 +128,132 @@ func main() {
 	c.Throttle(300, 100) // 300 hourly tokens, bursts of 100
 
 	for name, orgcfg := range cfg.Orgs {
-		if err := configureOrg(c, name, orgcfg); err != nil {
-			log.Fatalf("Configuration failed: %v", err)
+		if err := configureOrg(o, c, name, orgcfg); err != nil {
+			logrus.Fatalf("Configuration failed: %v", err)
 		}
 	}
 }
 
-func configureOrg(client *github.Client, orgName string, orgConfig org.Config) error {
+type configureOrgMembersClient interface {
+	BotName() (string, error)
+	ListOrgMembers(org, role string) ([]github.TeamMember, error)
+	RemoveOrgMembership(org, user string) error
+	UpdateOrgMembership(org, user string, admin bool) (*github.OrgMembership, error)
+}
+
+func configureOrgMembers(opt options, client configureOrgMembersClient, orgName string, orgConfig org.Config) error {
+	// Get desired state
+	wantAdmins := sets.NewString(orgConfig.Admins...)
+	wantMembers := sets.NewString(orgConfig.Members...)
+
+	// Sanity desired state
+	if n := len(wantAdmins); n < opt.minAdmins {
+		return fmt.Errorf("%s must specify at least %d admins, only found %d", orgName, opt.minAdmins, n)
+	}
+	var missing []string
+	for _, r := range opt.requiredAdmins.Strings() {
+		if !wantAdmins.Has(r) {
+			missing = append(missing, r)
+		}
+	}
+	if len(missing) > 0 {
```

Comment:
Not sure I fully understand this API -- for each org, the user provides the admins, a minimum amount of admins, and required global admins, then we check that the org-scoped list of admins is at least N long and explicitly requires all of the global ones? Why? Can't we just implicitly add the global (`opt.requiredAdmins`) list to the scoped lists?

Reply:
This is to mitigate the chance of disaster (losing access to the org from a bad config).

Making them flags (as opposed to part of the config) makes the failure domains more independent.

Thus if I now not only need to mess up the config, I also need to change the flags to be unsafe. Otherwise:
* we can fix the config and rerun the script because the bot is still an admin `--require-self`
* there will still be enough `--min-admins` add the bot back to the org, potentially explicitly including `--require-admins`
* we didn't delete enough of the admins to be a concern `--maximum-removal-delta`

The goal is to be able to say "regardless of whatever the config says ensure that `--require-admins=toc-member` is in the config's admin list. Otherwise refuse to apply it. Especially since it is easier to change the config than change flags.


See the Safety Checks and Disaster Recovery Plan in https://docs.google.com/document/d/18Kw8InxYk4OF14iTXXfvD2oy3AuoDxfqAmtvf6NbwBI/edit#heading=h.ohhjcvl2wm3y (and some commentary about it)